owncloudComLibrary/src/main/java/com/owncloud/android/lib/resources/shares/CreateRemoteShareOperation.kt

```diff
@@ -0,0 +1,191 @@
+/**
+ * Constructor
+ *
+ * @param remoteFilePath Full path of the file/folder being shared. Mandatory argument
+ * @param shareType      0 = user, 1 = group, 3 = Public link. Mandatory argument
+ * @param shareWith      User/group ID with who the file should be shared.  This is mandatory for shareType
+ * of 0 or 1
+ * @param publicUpload   If false (default) public cannot upload to a public shared folder.
+ * If true public can upload to a shared folder. Only available for public link shares
+ * @param password       Password to protect a public link share. Only available for public link shares
+ * @param permissions    1 - Read only Default for public shares
+ * 2 - Update
+ * 4 - Create
+ * 8 - Delete
+ * 16- Re-share
+ * 31- All above Default for private shares
+ * For user or group shares.
+ * To obtain combinations, add the desired values together.
+ * For instance, for Re-Share, delete, read, update, add 16+8+2+1 = 27.
+ */
+    (
+    private val remoteFilePath: String,
+    private val shareType: ShareType,
+    private val shareWith: String,
+    var publicUpload: Boolean?, // Upload permissions for the public link (only folders)
+    var password: String?, // Password to set for the public link
+    var permissions: Int // Access permissions for the file bound to the share
+) : RemoteOperation<ShareParserResult>() {
+    var getShareDetails: Boolean = false
+
+    /**
+     * Name to set for the public link
+     */
+    var name = ""
+
+    /**
+     * Expiration date to set for the public link
+     */
+    var expirationDateInMillis: Long = INITIAL_EXPIRATION_DATE_IN_MILLIS
+
+    init {
+        getShareDetails = false        // defaults to false for backwards compatibility
+    }
+
+    override fun run(client: OwnCloudClient): RemoteOperationResult<ShareParserResult> {
+        var result: RemoteOperationResult<ShareParserResult>
+
+        try {
+            val formBodyBuilder = FormBody.Builder()
+                .add(PARAM_PATH, remoteFilePath)
+                .add(PARAM_SHARE_TYPE, Integer.toString(shareType.value))
+                .add(PARAM_SHARE_WITH, shareWith)
+
+            if (name.isNotEmpty()) {
+                formBodyBuilder.add(PARAM_NAME, name)
+            }
+
+            if (expirationDateInMillis > INITIAL_EXPIRATION_DATE_IN_MILLIS) {
+                val dateFormat = SimpleDateFormat(FORMAT_EXPIRATION_DATE, Locale.getDefault())
+                val expirationDate = Calendar.getInstance()
+                expirationDate.timeInMillis = expirationDateInMillis
+                val formattedExpirationDate = dateFormat.format(expirationDate.time)
+                formBodyBuilder.add(PARAM_EXPIRATION_DATE, formattedExpirationDate)
+            }
+
+            if (publicUpload == true) {
+                formBodyBuilder.add(PARAM_PUBLIC_UPLOAD, publicUpload.toString())
+            }
+            if (!password.isNullOrEmpty()) {
+                formBodyBuilder.add(PARAM_PASSWORD, password!!)
```

Comment:
I think Android Studio suggests you to remove "!!", doesn't it?

Reply:
![screenshot 2019-03-05 at 18 13 44](https://user-images.githubusercontent.com/15261363/53823628-713aa080-3f72-11e9-89cb-dfaad8194f66.png)